cpp/include/cugraph/graph_functions.hpp

```diff
@@ -430,39 +455,63 @@ decompress_to_edgelist(
-                    std::optional<rmm::device_uvector<edge_time_t>>&& edgelist_edge_start_times,
-                    std::optional<rmm::device_uvector<edge_time_t>>&& edgelist_edge_end_times,
-                    bool reciprocal);
+           std::vector<arithmetic_device_uvector_t>>
+add_reciprocals_and_merge_edge_tuples(
+  raft::handle_t const& handle,
+  rmm::device_uvector<vertex_t>&& edgelist_srcs,
+  rmm::device_uvector<vertex_t>&& edgelist_dsts,
+  std::vector<arithmetic_device_uvector_t>&& edgelist_edge_properties,
+  std::vector<edge_key_selector_t> const& edge_key_selector,
+  std::vector<edge_value_combining_rule_t> const& edge_value_combining_rule,
+  bool store_transposed);
+
+/**
+ * @ingroup graph_functions_cpp
+ * @brief Remove non-reciprocal edges from the edgelist.  This was formerly called
+ * symmetrize_edgelist with reciprocal set to true.
+ *
+ * The input edge list scanned, each edge tuple that does not have a matching reciprocal edge is
+ * removed.  @p edge_key_selector is used to determine if an edge property is part of the uniqueness
+ * key of the edge.  If it is part of the uniqueness key, the edge tuple is removed if it does not
+ * have a matching reciprocal edge.  If it is not part of the uniqueness key, the edge property is
+ * ignored in the comparison.
+ *
+ * @tparam vertex_t Type of vertex identifiers. Needs to be an integral type.
+ * @tparam multi_gpu Flag indicating whether template instantiation should target single-GPU (false)
+ * or multi-GPU (true).
+ * @param handle RAFT handle object to encapsulate resources (e.g. CUDA stream, communicator, and
+ * handles to various CUDA libraries) to run graph algorithms.
+ * @param edgelist_srcs Vector of edge source vertex IDs. If multi-GPU, applying the
+ * compute_gpu_id_from_ext_edge_endpoints_t to every edge should return the local GPU ID for this
+ * function to work (edges should be pre-shuffled).
+ * @param edgelist_dsts Vector of edge destination vertex IDs.
+ * @param edgelist_edge_properties Vector of edge properties.
+ * @param edge_key_selector Vector of selectors for selecting the key of the edge properties.
+ * @param store_transposed Flag indicating whether to use sources (if false) or destinations (if
+ * true) as major indices in storing edges using a 2D sparse matrix.
+ * @return Tuple of non-reciprocal sources, destinations, and edge properties.
+ */
+template <typename vertex_t, bool multi_gpu>
+std::tuple<rmm::device_uvector<vertex_t>,
+           rmm::device_uvector<vertex_t>,
+           std::vector<arithmetic_device_uvector_t>>
+drop_nonreciprocal_edge_tuples(raft::handle_t const& handle,
+                               rmm::device_uvector<vertex_t>&& edgelist_srcs,
+                               rmm::device_uvector<vertex_t>&& edgelist_dsts,
+                               std::vector<arithmetic_device_uvector_t>&& edgelist_edge_properties,
+                               std::vector<edge_key_selector_t> const& edge_key_selector,
```

Comment:
We don't need `std::vector<edge_value_combining_rule_t> const& edge_value_combining_rule` here? If we have a reciprocal edges, are there no use-cases we need to combine edge properties? (e.g. update the edge weights to the average)

Reply:
Yes, I suppose we do.  I hadn't considered that case.  But if we have (1,2,1.0) and (2,1,2.0) then we would need to have the `edge_value_combining_rule_t` for that edge property to determine what value to assign to the pair of edges.  In that case `NONE` would leave the edges unmodified (and different in each direction), `ANY` would arbitrarily pick one of them to make them match.